File: src/api/subscriptions.ts
import { api } from './client'
import type { Subscription } from '@/shared/types'

export function fetchSubscriptions(): Promise<Subscription[]> {
  return api<Subscription[]>('/api/subscriptions')
}

export async function isSubscribed(channelId: string): Promise<boolean> {
  try {
    const { subscribed } = await api<{ subscribed: boolean }>(
      `/api/channels/${channelId}/subscribed`,
    )
    return subscribed
  } catch {
    return false
  }
}

export async function subscribe(channelId: string): Promise<void> {
  await api(`/api/channels/${channelId}/subscribe`, { method: 'POST' })
}

export async function unsubscribe(channelId: string): Promise<void> {
  await api(`/api/channels/${channelId}/subscribe`, { method: 'DELETE' })
}

File: backend/internal/handlers/subscriptions.go
package handlers

import (
	"net/http"

	"github.com/go-chi/chi/v5"
)

// GET /api/subscriptions
func (h *Handlers) ListSubscriptions(w http.ResponseWriter, r *http.Request) {
	subs, err := h.Store.ListSubscriptions(r.Context(), h.currentUserID(r))
	if handleStoreErr(w, err) {
		return
	}
	writeJSON(w, http.StatusOK, subs)
}

// GET /api/channels/{id}/subscribed
func (h *Handlers) IsSubscribed(w http.ResponseWriter, r *http.Request) {
	subscribed, err := h.Store.IsSubscribed(r.Context(), h.currentUserID(r), chi.URLParam(r, "id"))
	if handleStoreErr(w, err) {
		return
	}
	writeJSON(w, http.StatusOK, map[string]bool{"subscribed": subscribed})
}

// POST /api/channels/{id}/subscribe
func (h *Handlers) Subscribe(w http.ResponseWriter, r *http.Request) {
	if err := h.Store.Subscribe(r.Context(), h.currentUserID(r), chi.URLParam(r, "id")); handleStoreErr(w, err) {
		return
	}
	w.WriteHeader(http.StatusNoContent)
}

// DELETE /api/channels/{id}/subscribe
func (h *Handlers) Unsubscribe(w http.ResponseWriter, r *http.Request) {
	if err := h.Store.Unsubscribe(r.Context(), h.currentUserID(r), chi.URLParam(r, "id")); handleStoreErr(w, err) {
		return
	}
	w.WriteHeader(http.StatusNoContent)
}
